src/charm.py

```diff
@@ -9,93 +9,377 @@
+        if not self.model.config.get("database-name", None) and not self.model.config.get(
+            "topic-name", None
+        ):
+            event.fail("The database name or topic name is not specified in the config.")
             event.set_results({"ok": False})
             return
 
         mysql = self.mysql.fetch_relation_data()
         postgresql = self.postgresql.fetch_relation_data()
         mongodb = self.mongodb.fetch_relation_data()
+        kafka = self.kafka.fetch_relation_data()
 
-        if not mysql and not postgresql and not mongodb:
+        if not mysql and not postgresql and not mongodb and not kafka:
             event.fail("The action can be run only after relation is created.")
             event.set_results({"ok": False})
             return
 
         result = {"ok": True}
         if mysql:
-            result["mysql"] = list(mysql.values())[0]
+            result[MYSQL] = list(mysql.values())[0]
         if postgresql:
-            result["postgresql"] = list(postgresql.values())[0]
+            result[POSTGRESQL] = list(postgresql.values())[0]
         if mongodb:
-            result["mongodb"] = list(mongodb.values())[0]
+            result[MONGODB] = list(mongodb.values())[0]
+        if kafka:
+            result[KAFKA] = list(kafka.values())[0]
         event.set_results(result)
 
     def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
         """Event triggered when a database was created for this application."""
-        logger.info(f"database credentials are received: {event.username}")
-        self.unit.status = ActiveStatus(f"received {self.database} credentials")
+        logger.debug(f"database credentials are received: {event.username}")
+        self._on_config_changed(None)
+        if not self.unit.is_leader:
+            return
+        # update values in the databag
+        self._update_active_relations(event, "active")
+
+    def _on_topic_created(self, event: TopicCreatedEvent) -> None:
+        """Event triggered when a topic was created for this application."""
+        logger.debug(f"Kafka credentials are received: {event.username}")
+        self._on_config_changed(None)
+        if not self.unit.is_leader:
+            return
+        # update values in the databag
+        self._update_active_relations(event, "active")
+
+    def _update_active_relations(self, event: RelationEvent, status: str):
+        """Update the relation status in the peer-relation databag."""
+        self.set_secret("app", f"relation-{ event.relation.id}", status)
+
+    def _on_peer_relation_changed(self, _: RelationEvent):
+        """Handle the peer relation changed event."""
+        if not self.unit.is_leader:
+            return
+        removed_relations = []
+        # check for relation that has been removed
+        for k, v in self.app_peer_data.items():
+            if "relation-" in k:
+                if v == "broken":
+                    removed_relations.append(k)
```

Comment:
**chore:** Rename `k/v` to something descriptive. e.g `relation_data_key`, `value`